pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/performance/UseStringBufferForStringAppendsRule.java

```diff
@@ -39,92 +38,65 @@ public UseStringBufferForStringAppendsRule() {
+                || node.getNthParent(3) instanceof ASTForeachStatement) {
             return data;
         }
 
-        // Remember how often we the variable has been used
+        // Remember how often the variable has been used on the right hand side
         int usageCounter = 0;
 
-        for (NameOccurrence no : node.getUsages()) {
-            Node name = no.getLocation();
-            ASTStatementExpression statement = name.getFirstParentOfType(ASTStatementExpression.class);
-            if (statement == null) {
-                continue;
-            }
-            ASTArgumentList argList = name.getFirstParentOfType(ASTArgumentList.class);
-            if (argList != null && argList.getFirstParentOfType(ASTStatementExpression.class) == statement) {
-                // used in method call
-                continue;
-            }
-            ASTEqualityExpression equality = name.getFirstParentOfType(ASTEqualityExpression.class);
-            if (equality != null && equality.getFirstParentOfType(ASTStatementExpression.class) == statement) {
-                // used in condition
-                continue;
-            }
+        List<ASTNamedReferenceExpr> possibleViolations = new ArrayList<>();
+
+        for (ASTNamedReferenceExpr usage : node.getLocalUsages()) {
             if ((node.getNthParent(2) instanceof ASTFieldDeclaration
                     || node.getParent() instanceof ASTFormalParameter)
-                    && isNotWithinLoop(name)) {
+                    && isNotWithinLoop(usage)) {
                 // ignore if the field or formal parameter is *not* used within loops
                 continue;
             }
-            ASTConditionalExpression conditional = name.getFirstParentOfType(ASTConditionalExpression.class);
-
-            if (conditional != null) {
-                Node thirdParent = name.getNthParent(3);
-                Node fourthParent = name.getNthParent(4);
-                if ((Objects.equals(thirdParent, conditional) || Objects.equals(fourthParent, conditional))
-                        && conditional.getFirstParentOfType(ASTStatementExpression.class) == statement) {
-                    // is used in ternary as only option (not appended to other
-                    // string)
-                    continue;
+
+            boolean isSimpleAssignment = false;
+            if (usage.getParent() instanceof ASTAssignmentExpression) {
+                ASTAssignmentExpression assignment = (ASTAssignmentExpression) usage.getParent();
+                // it is either a compound (a += x)
+                if (assignment.isCompound()) {
+                    usageCounter++;
                 }
+
+                List<JVariableSymbol> symbolsOnTheRight = assignment.descendants(ASTInfixExpression.class)
+                        .filter(e -> e.getOperator() == BinaryOp.ADD)
+                        .children(ASTNamedReferenceExpr.class)
+                        .toList(ASTNamedReferenceExpr::getReferencedSym);
```

Comment:
I think we could avoid this intermediary list and just do something like

```java
int usageOnRightHandSide = 
  assignment.descendants(ASTInfixExpression.class)
            .filter(e -> e.getOperator() == BinaryOp.ADD)
            .children(ASTNamedReferenceExpr.class)
            .filter(ref -> node.getSymbol().equals(ref.getReferencedSym())
            .count();
```

Because `node.getSymbol()` is an alias of `usage.getReferencedSym()` in this scope.

Reply:
Actually we also should not recurse on the LHS, so we should do `assignment.getRightOperand().descendantsOrSelf(...)`